tensorflow/python/framework/tensor_util_test.py

```diff
@@ -226,13 +226,8 @@ def testHalf(self):
     t = tensor_util.make_tensor_proto(np.array([10.0, 20.0], dtype=np.float16))
     self.assertProtoEquals("""
       dtype: DT_HALF
-      tensor_shape {
-        dim {
-          size: 2
-        }
-      }
-      half_val: 18688
-      half_val: 19712
+      tensor_shape { dim { size: 2 } }
+      tensor_content: "\000I\000M"
```

Comment:
I don't like the change here. Can we keep the original? If not, why? Can we document what the binary data is supposed to be?

Reply:
Hi, @mihaimaruseac .

This is not an issue about correctness. It is an issue about performance. 

In tensorflow we have 2 ways to serialize tensor data in [protobuf](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.proto). 

- The first one is `repeated`. This method is to serialize data as many typed individual values. And it requires complex sterilization computation in dense tensor scenario.
- The second one is `tensor_content` which stores dense tensor data as unserialized binary. As the document says, "The purpose of this representation is to reduce serialization overhead during RPC call by avoiding serialization of many repeated small items."

In short, to achieve good performance, we should store dense tensor data in `tensor_content` in protobuf. However, the current API of `tf.make_tensor_proto` only store some very common types of data into `tensor_content`(not including `float16`).

In my case, I used the official ResNet50 model which uses `float16` tensors as input. And I got very expensive performance overhead in serialization(as what I said [here](https://github.com/tensorflow/tensorflow/pull/38258#issue-399389454)).

I think we'd better let `tf.make_tensor_proto` to store `float16` data in `tensor_content` rather than `repeated int32 half_val`. 

Because:
1. **Performance**. By using `tensor_content`, we can reduce 75% serialization overhead in RPC. In my case, from 280ms to 60ms, which is a very great performance improvement.
2. What's more. The origin `half_val` is also unreadable because it will regard the `float16` as `int`.
3. **Memory efficient**. `repeated int32 half_val` stores `float16` with 16 bits padding, while `tensor_content` just stores the raw layout.

Hence, I think this PR can bring better performance by default in the RPC scenario and won't cause any drawbacks.

I'm looking forward to your opinions.

Reply:
But this is only a test and we're changing the test from checking behavior when `repeated half_val` values are used to checking behavior when `float16` is used.

I think we should create a separate PR for the performance and enhancement of the proto and get that through a larger review, since it will need input from more people and multiple teams as well as changes throughout the base.

Reply:
Thanks for your reply. 

I modified the content of `_TENSOR_CONTENT_TYPES` and append it with `float16`. So I think it's not "only a test". And this will make `tf.make_tensor_proto` to store `float16` data in `tensor_content` rather than `repeated half_val`. As I changed the content of `_TENSOR_CONTENT_TYPES`, I also need to modify the test file, as the test result of `tf.make_tensor_proto` for `float16` was originally designed for `repeated half_val` other than `tensor_content`.

If you do not like this PR(though I think it only brings benefits), I'd like to recommend the users who met problems with serious serialization overhead to have a look at this PR. They can manually set the tensor buffer like `numpy.array.tostring()` into the `tensor_content` of tensor proto rather than using `tf.make_tensor_proto` who only set some common types into `tensor_content`. And be aware that the [official TensorFlow resnet50 NHWC model](https://github.com/tensorflow/models/tree/master/official/r1/resnet#pre-trained-model) used `float16` as input types and it will trigger the performance punishment if they use it with `tf.make_tensor_proto`.

Reply:
This is `testHalf`, it should test that `make_tensor_proto` works with a `np.float16`

There was a miscomunication here. Once you added the `float16` support the output of this test changed. Now it makes more sense.

Thank you for the patience